### PR TITLE
Add missing fields to the light sync state

### PR DESCRIPTION
**Pull request by expenses**
Part of #6804.

polkadot companion: https://github.com/paritytech/polkadot/pull/1801

### CODE REVIEW

File: client/chain-spec/src/chain_spec.rs [hunk 1]
```diff
@@ -400,7 +400,13 @@ where
 /// Hardcoded infomation that allows light clients to sync quickly.
 pub struct LightSyncState<Block: BlockT> {
 	/// The header of the best finalized block.
-	pub header: <Block as BlockT>::Header,
+	pub finalized_block_header: <Block as BlockT>::Header,
+	pub babe_finalized_block1_slot_number: sp_consensus_slots::SlotNumber,
+	pub babe_finalized_block_epoch_information: sc_consensus_babe::NextEpochDescriptor,
+	pub babe_finalized_next_epoch_transition: sc_consensus_babe::NextEpochDescriptor,
```
Comment (expenses) [inline client/chain-spec/src/chain_spec.rs:406]: @tomaka Do we also need `NextConfigDescriptor`s for these two fields?

File: client/service/src/chain_ops/build_sync_spec.rs [hunk 2]
```diff
@@ -14,23 +14,82 @@
...
+	let current_desc = babe_epoch_descriptors.next().unwrap()?;
 
 	Ok(sc_chain_spec::LightSyncState {
-		header
+		finalized_block_header: finalized_header,
+		babe_finalized_block1_slot_number: 0,
+		babe_finalized_block_epoch_information: current_desc,
+		babe_finalized_next_epoch_transition: next_desc,
+		grandpa_finalized_triggered_authorities: current_authorities,
+		grandpa_after_finalized_block_authorities_set_id: 0,
```
Comment (expenses) [inline client/service/src/chain_ops/build_sync_spec.rs:92]: We need a way to get the set id. Is calling into runtime for [`current_set_id`](https://substrate.dev/rustdocs/v2.0.0-rc6/pallet_grandpa/struct.Module.html#method.current_set_id) the best way to do this?

File: client/service/src/chain_ops/build_sync_spec.rs [hunk 3]
```diff
@@ -14,23 +14,82 @@
...
+			Err(err) => return Some(Some(Err(err))),
+		};
+
+		let change = header.digest()
+			.convert_first(|l| l.try_to(id).and_then(&func));
+
+		number -= NumberFor::<TBl>::one();
+		Some(change.map(Ok))
+	}).filter_map(|optional_log_item| optional_log_item)
+}
```
Comment (expenses) [inline client/service/src/chain_ops/build_sync_spec.rs:48]: This is a bit of a hack

File: client/service/src/chain_ops/build_sync_spec.rs [hunk 4]
```diff
@@ -14,23 +14,82 @@
...
+		}
+	);
+
+	let next_desc = babe_epoch_descriptors.next().unwrap()?;
+	let current_desc = babe_epoch_descriptors.next().unwrap()?;
 
 	Ok(sc_chain_spec::LightSyncState {
-		header
+		finalized_block_header: finalized_header,
+		babe_finalized_block1_slot_number: 0,
```
Comment (expenses) [inline client/service/src/chain_ops/build_sync_spec.rs:88]: Need a way to get this as well.

**Review by andresilva — COMMENTED** [review]
Not sure this is the way to go. I'll try to provide context on what the problem is and how we should go about fixing it.

Both BABE and GRANDPA maintain some data structures relating to the management of epochs (i.e. periods in which a given set of authorities is active). There's no need to go into details on the logic behind those data structures, but the key point is that they are populated as part of processing blocks. The data structures are:
- BABE - [EpochChanges](https://github.com/paritytech/substrate/blob/master/client/consensus/epochs/src/lib.rs#L300)
- GRANDPA - [AuthoritySet](https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/authorities.rs#L104)

In the scope of the light client, where we skip an initial sync and start from a block defined in the chainspec, this means that these data structures will not be populated with any data. As such, both [`BabeBlockImport`](https://github.com/paritytech/substrate/blob/master/client/consensus/babe/src/lib.rs#L1081) and [`GrandpaBlockImport`](https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/import.rs#L62) will fail to validate subsequent blocks (e.g. the errors you were seeing regarding BABE not finding epoch data). The solution is that we need to persist the data that exists in `EpochChanges` and `AuthoritySet` when we are creating the "light sync state", which we can then use when restoring from that state.

IMO as an initial implementation we can start by just dumping the data structures completely and persisting the whole thing (they are scale-encodable as we are already persisting them to the database). Once that is done we can prune these data structures to remove any unnecessary data (so that we don't persist everything). Both the data structures are easily available as part of service initialization: https://github.com/paritytech/substrate/blob/master/bin/node/cli/src/service.rs#L120 and https://github.com/paritytech/substrate/blob/master/bin/node/cli/src/service.rs#L112.

**Comment by andresilva** [comment]
We could persist all the data that is needed in separate fields (as it seems the intention in this PR so far), but we'd still need to populate `EpochChanges` and `AuthoritySet` with that data afterwards. Given that, I think it's easier to just persist `EpochChanges` and `AuthoritySet` to start with.

**Comment by expenses** [comment]
@andresilva done. The sync state increases the size of the chain spec by about 700kb, but that's not too bad that it's 3.8mb to begin with.

**Comment by expenses** [comment]
@andresilva How should we be getting the babe slot number for the first block? I thought that you'd be able to get this from the epoch changes, but you can't because it gets pruned.

**Comment by andresilva** [comment]
> @andresilva done. The sync state increases the size of the chain spec by about 700kb, but that's not too bad that it's 3.8mb to begin with.

Yeah, we'll have at least two lists of authority sets with public keys in there (BABE and GRANDPA). We might be able to prune some of the data you're seeing though (but let's worry about this later).

> @andresilva How should we be getting the babe slot number for the first block? I thought that you'd be able to get this from the epoch changes, but you can't because it gets pruned.

We can get it from BABE's runtime module https://github.com/paritytech/substrate/blob/master/frame/babe/src/lib.rs#L169. But do we need it? I know that Pierre mentioned this in the data we need to persist, but if we are persisting the latest epoch data I'm not sure we need this. At least the client code in substrate does not use this in any way, I believe this would only be needed in the `substrate-lite` implementation.

**Comment by expenses** [comment]
@andresilva we're trying to support substrate-lite here, but @tomaka if we don't use it in substrate, why is it needed in substrate-lite?

**Comment by expenses** [comment]
Also @tomaka am I right in thinking that `babe_finalized_block_epoch_information` is infomation about the epoch that contains the finalized header? We won't be able to get this from `EpochChanges` because it gets pruned once finalized.

**Comment by tomaka** [comment]
> We can get it from BABE's runtime module https://github.com/paritytech/substrate/blob/master/frame/babe/src/lib.rs#L169. But do we need it? I know that Pierre mentioned this in the data we need to persist, but if we are persisting the latest epoch data I'm not sure we need this.

I use it in substrate-lite to determine at which slot an epoch starts and ends.
The slot number of block 1 is the slot at which epoch 0 starts. Then epoch 1 starts at `slot_of_block_1 + slots_per_epoch`, and so on.
I guess that in Substrate the start slot number is stored in the epoch's information?

> Also @tomaka am I right in thinking that babe_finalized_block_epoch_information is infomation about the epoch that contains the finalized header?

Indeed. Sometimes we know that the finalized block is the last of the epoch, but not always.

**Comment by expenses** [comment]
> I guess that in Substrate the start slot number is stored in the epoch's information?

Yep: https://substrate.dev/rustdocs/v2.0.0-rc6/sc_consensus_babe/struct.Epoch.html#structfield.start_slot

**Comment by tomaka** [comment]
This is becoming off-topic, but I would need to learn a bit more about the long term plan concerning Babe slots. I know we're supposed to have some small NTP-like protocol, or something like that, but I don't know the details.

**Comment by tomaka** [comment]
Anyway, my point is that this PR should do the thing that makes sense, not the thing that brings compatibility with substrate-lite. It's substrate-lite which has to adjust.

**Comment by expenses** [comment]
Ok, I'm making progress getting the sync state to load in substrate (not lite). Now I'm getting:

```
Oct 01 15:10:15.402  WARN 💔 Error importing block 
0x3bf33ab015a8ae4ff09c8bc6e4361c82347cc87fefbba3478ee3b29cb0b71da1:
Err(Other(ClientImport("Parent block of 0x3bf3…1da1 has no associated weight")))
```

@andresilva  Should we saving babe block weights into the sync state as well?

**Comment by expenses** [comment]
As `0x3bf33ab015a8ae4ff09c8bc6e4361c82347cc87fefbba3478ee3b29cb0b71da1` is a child block of the finalized block here, `0xf355776a6aef623dc753ae505d467decb9cc2622d60ed5eb89bfbdf41f4bd0dd`, I think/hope I should only need to save its block weight.

**Comment by andresilva** [comment]
Yeah, I think just loading the weight of the finalized block from which we start syncing should be enough. We store the weight separately in aux storage (https://github.com/paritytech/substrate/blob/master/client/consensus/babe/src/aux_schema.rs#L129). Sorry I didn't remember about this beforehand. 😕

**Comment by tomaka** [comment]
For what it's worth, in substrate-lite whenever a new block arrives I find the common ancestor between this new block and the current best, and examine the path to determine which one has the higher weight. The weight itself, however, is never actually stored anyway. Only the relative weight between two blocks is ever calculated.

**Comment by expenses** [comment]
Ok, now what's happening is that we:

1. Import block 0xf355776a6aef623dc753ae505d467decb9cc2622d60ed5eb89bfbdf41f4bd0dd (number: 1,623,134).
2. Sync to block 0xa836daebc42349c356dfb3a4af9e8cd18add3f204483aa5f0d02c02548259fb2 (number: 1,623,272).
3. Panic with header not found in db: 0x6fd2538c07927ddacb9235a0cfc6cfdcfd92f2bac45f7fc7d4bbd43790181ab3 (number: 1,623,273).

I'm really not sure why it's trying to get the metadata of a block that it hasn't synced. Here's the relevant backtrace:

```

   7: <sc_client_db::light::LightStorage<Block> as sp_blockchain::header_metadata::HeaderMetadata<Block>>::header_metadata::{{closure}}::{{closure}}
             at client/db/src/light.rs:203
   8: core::option::Option<T>::ok_or_else
             at /home/ashley/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:545
   9: <sc_client_db::light::LightStorage<Block> as sp_blockchain::header_metadata::HeaderMetadata<Block>>::header_metadata::{{closure}}
             at client/db/src/light.rs:196
  10: core::option::Option<T>::map_or_else
             at /home/ashley/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:488
  11: <sc_client_db::light::LightStorage<Block> as sp_blockchain::header_metadata::HeaderMetadata<Block>>::header_metadata
             at client/db/src/light.rs:195
  12: <sc_light::blockchain::Blockchain<S> as sp_blockchain::header_metadata::HeaderMetadata<Block>>::header_metadata
             at client/light/src/blockchain.rs:95
  13: <sc_service::client::client::Client<B,E,Block,RA> as sp_blockchain::header_metadata::HeaderMetadata<Block>>::header_metadata
             at client/service/src/client/client.rs:1512
  14: sp_blockchain::header_metadata::lowest_common_ancestor
             at primitives/blockchain/src/header_metadata.rs:40
  15: sc_client_api::utils::is_descendent_of::{{closure}}
             at client/api/src/lib.rs:84
  16: <alloc::boxed::Box<F> as core::ops::function::Fn<A>>::call
             at /home/ashley/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1056
  17: fork_tree::node_implementation::Node<H,N,V>::find_node_index_where
             at utils/fork-tree/src/lib.rs:789
  18: fork_tree::node_implementation::Node<H,N,V>::find_node_index_where
             at utils/fork-tree/src/lib.rs:768
  19: fork_tree::node_implementation::Node<H,N,V>::find_node_index_where
             at utils/fork-tree/src/lib.rs:768
  20: fork_tree::node_implementation::Node<H,N,V>::find_node_index_where
             at utils/fork-tree/src/lib.rs:768
  21: fork_tree::node_implementation::Node<H,N,V>::find_node_where
             at utils/fork-tree/src/lib.rs:818
  22: fork_tree::ForkTree<H,N,V>::find_node_where
             at utils/fork-tree/src/lib.rs:279
  23: sc_consensus_epochs::EpochChanges<Hash,Number,E>::epoch_descriptor_for_child_of
             at client/consensus/epochs/src/lib.rs:577
  24: <sc_consensus_babe::BabeVerifier<Block,Client,SelectChain,CAW> as sp_consensus::import_queue::Verifier<Block>>::verify
             at client/consensus/babe/src/lib.rs:967
  25: sp_consensus::import_queue::import_single_block_metered
             at primitives/consensus/common/src/import_queue.rs:268
  26: sp_consensus::import_queue::basic_queue::import_many_blocks::{{closure}}
             at primitives/consensus/common/src/import_queue/basic_queue.rs:433
  27: <futures_util::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
             at /home/ashley/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/future/poll_fn.rs:54
  28: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
             at /home/ashley/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/future/future/map.rs:67
  29: <futures_util::future::future::Map<Fut,F> as core::future::future::Future>::poll
             at /home/ashley/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/lib.rs:107
  30: <futures_util::future::future::flatten::Flatten<Fut,<Fut as core::future::future::Future>::Output> as core::future::future::Future>::poll
             at /home/ashley/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/future/future/flatten.rs:47
  31: <futures_util::future::future::Then<Fut1,Fut2,F> as core::future::future::Future>::poll
             at /home/ashley/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/lib.rs:107
  32: sp_consensus::import_queue::basic_queue::BlockImportWorker<B,Transaction>::new::{{closure}}
             at primitives/consensus/common/src/import_queue/basic_queue.rs:223
```

**Comment by expenses** [comment]
The branch is https://github.com/paritytech/substrate/tree/ashley-improve-sync-state-WIP-loading. Run it with `rm -rf light && cargo run -- --light -d light`

**Comment by cheme** [comment]
Looking at the trace (failure on find lowest-common-ancestor), I would suspect something similar to the issue I had when doing #6851

**Comment by andresilva** [comment]
> I'm really not sure why it's trying to get the metadata of a block that it hasn't synced. Here's the relevant backtrace:

Most likely because that block is referenced in the `EpochChanges` struct. I suspect at block #1623272 we start a new epoch, and we need to do some queries in the `EpochChanges` tree to figure out which epoch data to use, in order to do that we need to perform some ancestry queries (i.e. is this new best block a descendent of a given node in the epoch tree?). So it should just be a matter of pruning "future" data from the epoch tree. But I'm not 100% sure about this and need to look into it so that I don't provide you with "half answers" anymore.

**Comment by expenses** [comment]
> > I'm really not sure why it's trying to get the metadata of a block that it hasn't synced. Here's the relevant backtrace:
> 
> Most likely because that block is referenced in the `EpochChanges` struct. I suspect at block #1623272 we start a new epoch, and we need to do some queries in the `EpochChanges` tree to figure out which epoch data to use, in order to do that we need to perform some ancestry queries (i.e. is this new best block a descendent of a given node in the epoch tree?). So it should just be a matter of pruning "future" data from the epoch tree (i.e. if our snapshot starts at block #10 our tree can't have epoch data for some epoch that was signalled at block #20). But I'm not 100% sure about this and need to look into it so that I don't provide you with "half answers" anymore.

That makes sense. When the sync state was created, the best block was near the head of the chain (1.7M ish), but the finalized block was way behind at (1623134).

**Comment by expenses** [comment]
So the solution here might be to trim the epoch changes in the sync state down, like we were going to do anyway.

**Comment by andresilva** [comment]
> So the solution here might be to trim the epoch changes in the sync state down, like we were going to do anyway.

Yeah, I think so, especially given what you said regarding finality lagging when the state was created (which means there's a high likelihood that the tree includes "future" epochs). I have tested your branch and made some changes (https://github.com/paritytech/substrate/tree/andre/filter-babe-epoch-changes-tree). The first commit is some example code (very dirty WIP) to prune the tree out of any "future" epoch data, the node should print the tree before and after on startup.

The second commit just comments some code that prunes the tree on finality. The reason is that this is also triggering some "header not found" errors. The reason being that the tree also includes "old" epoch data for which we don't have the headers (I still don't know what is the best way of fixing this, will get back to you soon).

Either way, with those changes the node successfully syncs to the tip, so we're really close! 😀🎉

**Comment by expenses** [comment]
@andresilva awesome!

**Comment by expenses** [comment]
By filtering the fork tree before serializing, the sync state only adds about 100kb to the chain spec.

**Comment by expenses** [comment]
In https://github.com/paritytech/substrate/pull/6894, I originally started out having a RPC call to generate a chain spec with a sync state, but changed it to a CLI subcommand as it seemed easier than starting a light client, getting it to sync, and making the RPC call. Now that we're generating the sync state on a full client though, I'd like to return to having an RPC call for this. It's a lot easier to test, less fiddly and would allow us to do stuff like generate the sync states on an archive node periodically via a cron job.

@bkchr what do you think?

**Comment by bkchr** [comment]
Would the light client query the full node to get the latest sync state?
Or would this still work by using the chain spec?

**Comment by expenses** [comment]
> Would the light client query the full node to get the latest sync state?
> Or would this still work by using the chain spec?

This would all be done via chain specs.

**Comment by bkchr** [comment]
Why would that all be done via chain specs? Assuming that a light client always connects to a full node, why does this data even need to be part of the chain spec?

**Comment by tomaka** [comment]
Well, trust. You trust the chain specs, but you can't know if the full node is lying to you.

**Comment by bkchr** [comment]
Yeah good point. We can add the RPC if you need it, but please make the RPC unsafe.

However I already see people having "trusted" RPC nodes which they will query for the latest chain spec.

File: client/rpc-api/src/system/error.rs [hunk 5]
```diff
@@ -32,6 +32,12 @@ pub enum Error {
 	NotHealthy(Health),
 	/// Peer argument is malformatted.
 	MalformattedPeerArg(String),
+	/// An error that occurred while trying to generate the sync state.
+	#[display(fmt = "{}. Try syncing the node some more.", _0)]
+	GenSyncState(sp_blockchain::Error),
+	/// BuildNetworkParams::sync_state_items has not been set.
+	#[display(fmt = "BuildNetworkParams::sync_state_items has not been set.")]
+	MissingSyncStateItems,
```
Comment (expenses) [inline client/rpc-api/src/system/error.rs:40]: There might be a better way to do these errors.

File: client/service/src/builder.rs [hunk 6]
```diff
@@ -780,6 +780,12 @@ pub struct BuildNetworkParams<'a, TBl: BlockT, TExPool, TImpQu, TCl> {
 	pub finality_proof_request_builder: Option<BoxFinalityProofRequestBuilder<TBl>>,
 	/// An optional, shared finality proof request provider.
 	pub finality_proof_provider: Option<Arc<dyn FinalityProofProvider<TBl>>>,
+	/// Shared GRANDPA and BABE items that are used to generate sync states to be used by light
+	/// clients. Only generated on full clients.
+	pub sync_state_items: Option<(
+		grandpa::SharedAuthoritySet<<TBl as BlockT>::Hash, NumberFor<TBl>>,
+		sc_consensus_epochs::SharedEpochChanges<TBl, sc_consensus_babe::Epoch>,
+	)>
```
Comment (expenses) [inline client/service/src/builder.rs:788]: The name of this param could possibly be improved.

**Comment by expenses** [comment]
In what situations are there no pending changes in the authority set? calling `pending_changes().next()` was working fine on the previous version of flaming fir, but not on the current version or polkadot.

**Comment by andresilva** [comment]
> In what situations are there no pending changes in the authority set? calling `pending_changes().next()` was working fine on the previous version of flaming fir, but not on the current version or polkadot.

Under normal / healthy conditions there should never be any pending changes, or at most 1 and only for a brief period. Pending authority set changes are enacted whenever the block in which they were triggered is finalized, if finality is working properly after the change is announced that block should get finalized quickly. Either finality was not working on flaming fir before the reset, or the node was not getting the latest finality data (which is kind of expected in the current state if it was a light client).

**Comment by expenses** [comment]
Ah, okay. So I guess we can change `grandpa_finalized_scheduled_change` to be an option then.

**Comment by andresilva** [comment]
IMO we should be persisting the whole `AuthoritySet` data structure as we do for `EpochChanges`. What I said above only applies to the "happy"/normal path :P In reality in the `AuthoritySet` we still keep a tree structure (that uses `fork-tree` as well but has slightly different semantics) for the cases where we have more than one pending change (which could potentially be on different forks). 

An alternative is to only store one pending change as is done now but only allow generating such a sync state if there is only one or zero pending changes in the `AuthoritySet`, i.e. if finality was halted on a network and there multiple pending changes the RPC would refuse generating a sync state.

**Comment by tomaka** [comment]
I thought that there could only be one pending change at a time on the finalized chain?

**Comment by expenses** [comment]
Ok, I think this should be ready now. We can save things like filtering the epoch changes for the next PR.

**Comment by expenses** [comment]
I'm getting a new panic on [`ashley-improve-sync-state-WIP-loading`](https://github.com/paritytech/substrate/compare/ashley-improve-sync-state...ashley-improve-sync-state-WIP-loading) that I wasn't getting with the previous version of flaming fir.

File: client/chain-spec/src/lib.rs [hunk 7]
```diff
@@ -149,6 +149,8 @@ pub trait ChainSpec: BuildStorage + Send {
 	fn add_boot_node(&mut self, addr: MultiaddrWithPeerId);
 	/// Return spec as JSON.
 	fn as_json(&self, raw: bool) -> Result<String, String>;
+	/// Return spec as JSON value.
+	fn as_json_value(&self, raw: bool) -> Result<serde_json::Value, String>;
```
Comment (bkchr) [inline client/chain-spec/src/lib.rs:153]: Can we not change the function above?  I don't see any reason why we need this two times. And yes, I see that one is a String and the other one a value.

File: client/service/src/builder.rs [hunk 8]
```diff
@@ -813,6 +813,12 @@ pub struct BuildNetworkParams<'a, TBl: BlockT, TExPool, TImpQu, TCl> {
 	pub finality_proof_request_builder: Option<BoxFinalityProofRequestBuilder<TBl>>,
 	/// An optional, shared finality proof request provider.
 	pub finality_proof_provider: Option<Arc<dyn FinalityProofProvider<TBl>>>,
+	/// Shared GRANDPA and BABE items that are used to generate sync states to be used by light
```
Comment (bkchr) [inline client/service/src/builder.rs:816]: What is when someone isn't using BABE or GRANDPA? Can't this be an optional RPC that we don't hard code here?

Comment (expenses) [reply client/service/src/builder.rs]: Done

File: client/chain-spec/src/lib.rs (hunk 7 above)
Comment (expenses) [reply client/chain-spec/src/lib.rs:153]: Making this change causes `build_spec_works` to fail, because of `Error: Service(Other("Error generating spec json: u128 is not supported"))`. I think this is because `serde_json::to_value` panics on `u128`s while `serde_json::to_string_pretty` does not seem to.

This is a gotcha that we need to look out for in the RPC call too.

Comment (bkchr) [reply client/chain-spec/src/lib.rs:153]: Where is the `u128` coming from? 

We already use this "hack" in transaction payment: https://github.com/paritytech/substrate/blob/master/frame/transaction-payment/rpc/runtime-api/src/lib.rs#L40-L43

Comment (expenses) [reply client/chain-spec/src/lib.rs:153]: I'm not sure. This only happens when running `substrate build-spec --dev`, but not without `--dev`. The balance type for the node is `u128`, so that could be why?

Comment (bkchr) [reply client/chain-spec/src/lib.rs:153]: Yeah sounds right, but we still should not have 2 functions which are doing the same. If you need a serde::Value for the RPC, you could still deserialize the String to a `Value`. I know this is also shit, but we should not blow up our code base because Serde isn't able to handle this correctly...

**Comment by expenses** [comment]
@andresilva on `ashley-improve-sync-state-WIP-loading` im not getting a lot of this warning:

`2020-10-12 14:25:18  💔 Error importing block 0x8ce0f445b3f00a3585b5571492cbda40bc04bb09f64179e0723628a6bf9448a7: Err(Other(ClientImport("Unexpected epoch change")))`

**Comment by expenses** [comment]
Ok, so I've updated `ashley-improve-sync-state-WIP-loading` to not generate CHTs in `LightStorage::note_finalized` and it now syncs the provided chain spec with a sync state for block `#118347` to the head without panicking.

**Comment by expenses** [comment]
It is still panicking on a second chain spec on branch `ashley-improve-sync-state-WIP-loading-panicking` that from block `#198161`.

**Comment by expenses** [comment]
This PR is ready and I'd like to get it merged soon. The two things to check in reviews are:

1. Are we completely sure that the sync state contains everything we need?
2. Is the way I've set up this RPC call correct?

**Review by andresilva — APPROVED** [review]
just a minor nit but overall lgtm.

**Comment by andresilva** [comment]
> I thought that there could only be one pending change at a time on the finalized chain?

There can be multiple pending changes on the **unfinalized** chain (if GRANDPA has stalled). On Kusama we will be rotating the session every hour, which potentially leads to an authority set change in GRANDPA. Since the changes are only enacted on finality, if GRANDPA stops working for 6h we can end up with 6 pending authority set changes all on the same branch (which we need to finalize in-order afterwards).

File: client/finality-grandpa/src/authorities.rs [hunk 9]
```diff
@@ -53,7 +53,7 @@ impl<H, N> Clone for SharedAuthoritySet<H, N> {
 
 impl<H, N> SharedAuthoritySet<H, N> {
 	/// Acquire a reference to the inner read-write lock.
-	pub(crate) fn inner(&self) -> &RwLock<AuthoritySet<H, N>> {
+	pub fn inner(&self) -> &RwLock<AuthoritySet<H, N>> {
```
[the remark this answers is not in the document]
Comment (expenses) [reply client/finality-grandpa/src/authorities.rs]: :+1:

**Comment by tomaka** [comment]
> on the unfinalized chain

But this PR is only about returning the state of the finalized chain, no? A data structure containing the state of the finalized chain doesn't have to contain more than one pending change.

**Comment by expenses** [comment]
@tomaka Well no, at the moment we just encode the whole `AuthoritySet` at the time of the RPC call. If finality is lagging behind significantly, then it will contain unfinalized pending changes. In the future we could trim this down.

**Comment by expenses** [comment]
Ok, I'd like to get another approval and then I'll merge this.

**Comment by andresilva** [comment]
> @tomaka Well no, at the moment we just encode the whole `AuthoritySet` at the time of the RPC call. If finality is lagging behind significantly, then it will contain unfinalized pending changes. In the future we could trim this down.

To get something working I think it was easier to just persist BABE's `EpochChanges` and GRANDPA's `AuthoritySet` data structures. This is because we can then just blindly encode / decode and have something working. This has some disadvantages though: it includes unnecessary data and it's harder to integrate with other clients (e.g. substrate-lite doesn't use these data structures so cannot easily use these fields). I think the next step should be to trim them down to the bare essentials and in which case we will probably end up only persisting those pieces of data and not the complete data structure (e.g. for GRANDPA it would be the set id, current authorities and some pending standard change that may exist). But IMO this can be done in a follow-up PR.

tldr: the actual sync state data will probably change after this PR, IMO this is fine as we can get something working in the meantime without having to deal with all the details of what needs to be persisted.

**Comment by expenses** [comment]
>tldr: the actual sync state data will probably change after this PR, IMO this is fine as we can get something working in the meantime without having to deal with all the details of what needs to be persisted.

As long as the types don't change, I'm okay with this.

**Comment by expenses** [comment]
bot merge

**Comment by andresilva** [comment]
> > tldr: the actual sync state data will probably change after this PR, IMO this is fine as we can get something working in the meantime without having to deal with all the details of what needs to be persisted.
> 
> As long as the types don't change, I'm okay with this.

If we want to make it easier to be consumed by `substrate-lite` then we probably want to change the types (to "plain" data as you had initially). TBH I wasn't thinking about interoperability when I suggested persisting `EpochChanges` and `AuthoritySet` just thinking of the easiest / fastest way to get the sync state to work. But IMO if we're going to make the substrate light client work first with the sync state (and look into substrate-lite afterwards) then the current approach is OK since there are some unknowns and we will probably only be sure once we have the whole thing working properly.

File: client/sync-state-rpc/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+		)?
+			.ok_or_else(|| sp_blockchain::Error::Msg(
+				format!("Failed to load the block weight for block {:?}", finalized_hash)
+			))?;
+
+		Ok(sc_chain_spec::LightSyncState {
+			finalized_block_header: finalized_header,
+			babe_epoch_changes: self.shared_epoch_changes.lock().clone(),
+			babe_finalized_block_weight: finalized_block_weight,
+			grandpa_authority_set: self.shared_authority_set.clone_inner(),
```
Comment (cheme) [inline client/sync-state-rpc/src/lib.rs:98]: can this be prune (grandpa set or babe epochs) concurrently ?
btw even if possible the race seems highly improbable (probably when code get generic there will be 
a trait method like `synch_state_at(block_hash) -> Option<S>` for substrate component that require synch state and this will not be a possibillity.

Comment (andresilva) [reply client/sync-state-rpc/src/lib.rs:98]: It should be fine since both BABE and GRANDPA will acquire a lock when writing to these data structures (e.g. if there's a block being imported).

Comment (cheme) [reply client/sync-state-rpc/src/lib.rs:98]: so maybe we should lock both structure before getting the finalized block hash? (it would make thing easier to read at least)
